pkg/apis/intevents/v1/pullsubscription_defaults.go

```diff
@@ -0,0 +1,60 @@
+/*
+Copyright 2020 Google LLC.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"context"
+	"time"
+
+	"knative.dev/pkg/apis"
+
+	duck "github.com/google/knative-gcp/pkg/apis/duck"
+	"knative.dev/pkg/ptr"
+)
+
+const (
+	defaultRetentionDuration = 7 * 24 * time.Hour
+	defaultAckDeadline       = 30 * time.Second
+)
+
+func (s *PullSubscription) SetDefaults(ctx context.Context) {
+	ctx = apis.WithinParent(ctx, s.ObjectMeta)
+	s.Spec.SetDefaults(ctx)
+	duck.SetAutoscalingAnnotationsDefaults(ctx, &s.ObjectMeta)
+}
+
+func (ss *PullSubscriptionSpec) SetDefaults(ctx context.Context) {
+	if ss.AckDeadline == nil {
+		ackDeadline := defaultAckDeadline
+		ss.AckDeadline = ptr.String(ackDeadline.String())
+	}
+
+	if ss.RetentionDuration == nil {
+		retentionDuration := defaultRetentionDuration
+		ss.RetentionDuration = ptr.String(retentionDuration.String())
+	}
+
+	ss.PubSubSpec.SetPubSubDefaults(ctx)
+
+	switch ss.Mode {
+	case ModeCloudEventsBinary, ModeCloudEventsStructured, ModePushCompatible:
+		// Valid Mode.
+	default:
+		// Default is CloudEvents Binary Mode.
+		ss.Mode = ModeCloudEventsBinary
+	}
```

Comment:
I think we should actually remove Mode from PullSubscription...
If you want to do it in a follow up, that's fine, just create an issue so that we tackle this just after this one gets merged

Reply:
given the flakiness that exists in the CI, it might be better doing it now actually